turtle_tf2_cpp/src/turtle_tf2_listener.cpp

```diff
@@ -88,42 +66,92 @@ class FrameListener : public rclcpp::Node
     std::string fromFrameRel = target_frame_.c_str();
     std::string toFrameRel = "turtle2";
 
-    geometry_msgs::msg::TransformStamped transformStamped;
-
-    // Look up for the transformation between target_frame and turtle2 frames
-    // and send velocity commands for turtle2 to reach target_frame
-    try {
-      transformStamped = tf_buffer_->lookupTransform(
-        toFrameRel, fromFrameRel,
-        tf2::TimePoint(),
-        500ms);
-    } catch (tf2::LookupException & ex) {
-      RCLCPP_INFO(this->get_logger(), "transform not ready");
-      return;
+    if (!turtle_spawned_) {
+      // Check if the service is available
+      while (!spawner_->wait_for_service(1s)) {
+        if (!rclcpp::ok()) {
+          RCLCPP_ERROR(
+            this->get_logger(),
+            "Interrupted while waiting for the service. Exiting."
+          );
+          timer_->cancel();
+          return;
+        }
+        RCLCPP_INFO(
+          this->get_logger(),
+          "Service not available, waiting again..."
+        );
+      }
```

Comment:
The issue with this is that the timer expires every second, but we are waiting for up to a second for the service to become available.  What that means is that if the service isn't available, we'll wait for the second, and then the timer will expire and we'll immediately try again.

What I think you should do instead here is just *check* if the service is available every time the timer expires, but not wait for it.  That is, the code should look something like:

```
  if (!spawner_->service_is_ready()) {
    return;
  }
```

Then you don't have to cancel the timer, you don't have to check for `rclcpp::ok`, etc.

Reply:
Solved in https://github.com/ros/geometry_tutorials/pull/53/commits/4a380e7cc5b293228d224a429d4d5fbce7a168db